Task9.2

bekhhh opened the pull request:
вывел в фрпмате json, но не факт что правильно

bekhhh committed:
подсказки не помешали бы
вывел в фрпмате json, но не факт что правильно

usik-vladislav commented on Task9/Task9/Character.cs:
```diff
@@ -32,5 +34,11 @@ public void PrintItems()
+
+        public void PrintJson()
+        {
+            var json = JsonConvert.SerializeObject(this, Newtonsoft.Json.Formatting.Indented);
```
Cейчас выводит среди полей UniqueCharacteristics. Его по условию не нужно выводить.
Почитай как в библиотеке Newtonsoft.Json игнорировать поля для записи в json. Возможно понадобится почитать про атрибуты в C#.

usik-vladislav commented on Task9/Task9/InputHandler.cs:
```diff
@@ -31,17 +32,21 @@ public void HandlerInput()
                 {
-                    Console.WriteLine("Введены неверные данные");
-                    continue;
+                    if (characters.TryGetValue(words[1], out character)) 
+                    {
+                        Console.WriteLine($"You choose {words[1]} ");
+                        character.PrintJson();
+                        continue;
+                    }
                 }
                 if (words[0] != "get")
                 {
                     Console.WriteLine("Введены неверные данные");
                     continue;
                 }
-                if (!characters.TryGetValue(words[2], out var character))
+                if (!characters.TryGetValue(words[2], out character))
                 {
                     Console.WriteLine("Такого персонажа не существует");
                     continue;
```
Если написать get items будет ошибка

usik-vladislav committed:
Changed 7 task description

bekhhh committed:
мне кажется ошибок минимум, но никогда не равна нулю
мало проверок, тк мы не делали на них акцент в этой задаче

bekhhh committed:
Merge branch 'Task9.2' of https://github.com/bekhhh/learning into Task9.2

bekhhh committed:
Merge branch 'master' into Task9.2

# Conflicts:
#	Tasks.txt

usik-vladislav commented on Task9/Task9/Elements.cs:
```diff
@@ -0,0 +1,10 @@
+        Fire,
+        Water,
+        Air
+    }
```
Чаще всего различные классы, структуры, енамы и т д называют в единственном числе, так что лучше здесь просто Element. Конкретно к енамам иногда добавляют в конце Type, типа ElementType (Тип элемента). Но имхо лучше просто Element.
!!! Когда переименуешь класс, не забудь переименовать файл (если студия не сделает это автоматически)

Также файл нужно перенести из корня проекта в какую-нибудь осмысленную папку. Например создать в Characters папку Ability и туда положить файлы Ability и Element.

usik-vladislav commented on Task9/Task9/Ability.cs:
```diff
@@ -0,0 +1,17 @@
+            Mana = mana;
+            Elements = elements;
+        }
+    }
```
1.У тебя у способности ОДИН элемент, так что и название свойства должно быть в единственном числе -- Element.
2.Файл нужно убрать из корня проекта.
3.Сейчас поле Elements выводится в json как 0, 1, 2, 3 -- например
 {
    "Name": "Залп",
    "Description": "Крутая",
    "Mana": 5,
    "Elements": 0
  }.
  Можно сделать так, чтобы Енам выводилось в Json в виде строки, чтобы было "Elements": "Earth".
  В разных библиотеках это делается по разному. Загугли как это делается в newtonsoft и сделай вывод Element строкой.

usik-vladislav commented on Task9/Task9/InputHandler.cs:
```diff
@@ -8,51 +10,96 @@ public class InputHandler
+           { nameof(Rogue), new Rogue() },
+           { nameof(Wizard), new Wizard() },
+           { nameof(Warrior), new Warrior() }
+        };
```
Поправить отступы

usik-vladislav commented on Task9/Task9/Character.cs:
```diff
@@ -1,36 +1,54 @@
+            var abilitiesJson = JsonConvert.SerializeObject(Abilities, Newtonsoft.Json.Formatting.Indented);
+            Console.WriteLine(json);
+            Console.WriteLine("Abilities:");
+            Console.WriteLine(abilitiesJson);
```
Способности не должны выводиться отдельно от персонажа, они должны быть внутри скобок (json) описания персонажа (см условие). Тут достаточно просто убрать JsonIgnore и выводить персонажа в json вместе со способностями, то есть по сути тебе нужно PrintJson а не новый метод.

usik-vladislav commented on Task9/Task9/InputHandler.cs:
```diff
@@ -8,51 +10,96 @@ public class InputHandler
+                    {
+                        if (words[0] == "start")
+                        {
+                            if (characters.TryGetValue(words[1], out character))
+                            {
+                                Console.WriteLine($"You choose {words[1]} ");
+                                character.PrintJson();
+                                hasStarted = true;
+                                continue;
+                            }
+                        }
+                        else
+                        {
+                            Console.WriteLine("Введены неверные данные");
+                        }
+                    }
+                    else
+                    {
+                        if (words[0] == "add" && words[1] == "ability")
+                        {
+                            if (!Enum.TryParse(words[5], out Elements elements))
+                            {
+                                Console.WriteLine("Введены неверные данные");
+                                continue;
+                            }
+                            var ability = new Ability(words[2], words[3], int.Parse(words[4]), elements);
+                            character.Abilities.Add(ability);
+                            Console.WriteLine($"Добавлена способность {ability.Name}");
+                            continue;
+                        }
+                        if (words[0] == "show" && words[1] == "info")
+                        {
+                            character.PrintInfo();
+                            continue;
+                        }
+                        else
+                        {
+                            Console.WriteLine("Введены неверные данные");
+                        }
+                    }
```
Класс InputHandler очень разросся и метод стал большим и сложным для понимания. Будем его рефакторить. 
Я тебе напишу примерную инструкцию, как можно улучшить код, твоя задача ее выполнить =).
По итогу я хочу, чтобы
а) Вся логика обработки строки пользователя была вынесена из класса InputHandler
b) Обработка строки пользователя была более четкой и не допускала ошибок (сейчас при неправильном вводе зачастую в консоль выводится непонятная ошибка на английском, а надо понятное сообщение на русском)
с) При различных неправильных вводах выводилось понятное сообщение о том, что надо исправить. 
d) В консоль на нужных этапах выводилась инструкция по формату ввода.
Разобьем задачу на отдельные пункты.
1.Отделим логику парсинга строки и определения действия (что нужно сделать) от исполнения действия (вывод предметов, персонажа). Для этого сделаем отдельный енам Command, который будет означать конкретную комманду -- GetDescription, GetItems, Start, AddAbility, ShowInfo, InvalidInput. Также создадим отдельные классы CommandParser и CommandParsingResult. Первый класс будет содержать один метод Parse принимающий строку input и возвращающий CommandParsingResult. Второй класс просто будет содержать поля Command и string Message.
2. Реализуем парсинг (метод Parse). По сути это будет большое количество различных ветвистых проверок по результату которых мы отдаем конкретную команду и сообщение для вывода в консоль (при надобности). Если в какой-то момент логики понимаем, что строка неправильная, возвращаем результат с Command.InvalidInput и Message с описанием, что именно неправильно. Ниже полная инструкция по порядку.
- Проверяем на налл и пустую строку, при ошибке Message - Не получилось распознать команду. Введите строку в нужном формате.
- если первое слово get провереям
 -- длина ровно 3 слова, при ошибке Message (далее по тексту буду писать просто Message) -- Команда get состоит 3 слов, введите команду в нужном формате.
 -- второе слово description или items. Message -- Не получилось распознать команду get (введенное второе слово). Введите команду в нужном формате.
 -- третье слово из списка персонажей. Message -- Персонажа (третье слово) не существует. Укажите в команде персонажа из начального списка.
 -- если все проверки прошли успешно возвращаем Command.GetDescrtiption или Command.GetItems (В зависимости от 2 слова) и Message = null.
 (далее буду писать Message только в неочевидных местах, но сам Message должен быть во всех местах с неправильным вводом, как было выше. Сам Message пиши по аналогии с примерами выше.)
- если первое слово start
 -- длина ровно 2 слова
 -- второе слово из списка персонажей
 -- до текущего момента в программе не выбиралась команда start. Message -- Команду start можно использовать один раз за запуск. Введите другую команду.
 -- все проверки успешны. Command.Start и Message = null
- если первое слово show
 -- длина ровно 2 слова
 -- второе слово show. Message -- не удалось распознать команду show (второе слово)
 -- все успешно -- Command.ShowInfo
- если первое слово add
 -- длина ровно 5 слов
 -- второе слова ability
 -- 3 слово не содержит чисел и знаков препинания. Message -- Имя способности не может содержать числа и знаки препинания. Выберете другое имя для способности.
 -- 4 слово является числом. Message -- В команде add ability на 4 позиции должно быть число, указывающее стоимость в мане.
 -- 5 слово является одним из Enum. Message -- В команде add ability на 5 позиции должно быть слово одной из стихии: Earth, Fire, Water, Air.
 -- Все успешно -- Comand.AddAbility.
- если 1 слово не попало ни в одно ветвление выше, то ошибка. Message -- Не удалось распознать команду (первое слово). Введите команду из инструкции.
3.Меняем код в InputHandler. 
 - Создаем свойство для хранения CommandParser и инициализируем его в конструкторе.
 - Удаляем внутри try все кроме Console.ReadLine.
 - Передаем в парсер input и получаем в var result результат парсинга.
 - Далаем Switch по result.Command
  -- Если result.Command == Command.InvalidInput, то выводим result.Message в консоль и делаем Continue.
  -- если result.Command == Command.GetItems, то выводим предметы.
  -- и т. д.
4. Добавляем вывод инструкций.
- Создаем отдельный статический класс для хранения строк инструкций, у него делаем два константных поля StartRules и AfterStartRules.
- StartRules -- Напиши текст описывающий какие команды доступны -- get items|description, start, так чтобы сторонний человек понял, что нужно вводить, чтобы программа работала. Также опиши, что start character может быть вызвана только один раз.
- AfterStartRules -- опиши новые команды add ability и show info, аналогично StartRules.
- В InputHandler до начала цикла при старте выводишь в консоль StartRules.
- После выполнения команды start выводишь AfterStartRules

bekhhh committed:
переименовал энамку

bekhhh committed:
dfs

bekhhh committed:
не могу найти что не так, пока что остановился тут)

bekhhh committed:
add else if

bekhhh committed:
в swithc добавил showInfo (изначально забыл)
теперь команду start можно использовать только один раз
в команде add ability поправил пару критических моментов как проверка длины и наличие описания способности(в начале пропустил этот момент) )

bekhhh committed:
add поле в класс Parser

bekhhh committed:
df

bekhhh committed:
qwe

bekhhh committed:
add UnitTests

bekhhh committed:
Parse CommandParser

usik-vladislav commented on Task9/Task9/UserInput/CommandParsing/CommandParser.cs:
```diff
@@ -8,54 +9,80 @@ public class CommandParser
                     .ToArray();
-            if (words[0] == "get")
+
+            switch (_words[0])
```
Упадет при обращении по индексу, если передать строку из пробелов

usik-vladislav commented on Task9/Task9/UserInput/CommandParsing/CommandParser.cs:
```diff
@@ -8,54 +9,80 @@ public class CommandParser
+        {
+            if (_words[0] == "get")
+            {
+                if (_words.Length != 3)
```
Нужно использовать input вместо _words, раз используешь параметр. И вообще лучше отказаться от поля _words и передавать данные только через параметр. Передача данных между методами с помощью поля плохая практика

usik-vladislav commented on Task9/Task9/UserInput/CommandParsing/CommandParser.cs:
```diff
@@ -8,54 +9,80 @@ public class CommandParser
                 }
-                if (!characters.TryGetValue(words[2], out character))
+                if (!_characters.TryGetValue(_words[2], out character))
                 {
```
Записывай персонажа в локальную переменную метода. Нет надобности использовать поле.

usik-vladislav commented on Task9/Task9/UserInput/CommandParsing/CommandParser.cs:
```diff
@@ -66,59 +93,68 @@ public CommandParsingResult Parse(string input)
-            "Не удалось распознать команду (первое слово). Введите команду из инструкции.");
+
+            _newAbility = new Ability(input[2], input[3], _manaCost, _element);
+            return new CommandParsingResult(Command.AddAbility, ability: _newAbility, character: character);
```
Не нужно здесь возвращать персонажа. Сделай поле персонажа в классе InputHandler и сохраняй персонажа из результата туда (если персонаж не равен null).

usik-vladislav commented on Task9/Task9/UserInput/CommandParsing/CommandParser.cs:
```diff
@@ -8,54 +9,80 @@ public class CommandParser
+                    return ParseAddAbility(_words);
+                default:
+                    return new CommandParsingResult(Command.InvalidInput,
+                    "Не удалось распознать команду (первое слово). Введите команду из инструкции.");
+            }
+        }
```
Когда я тебе писал в описании (первое слово), (второе слово) и т д, я подразумевал подставленное значение из words

bekhhh committed:
refactor code

usik-vladislav commented on Task9/Task9/UserInput/InputHandler.cs:
```diff
@@ -0,0 +1,61 @@
+                {
+                    var input = Console.ReadLine();
+                    var result = Parser.Parse(input);
+                    if (result.Character is not null) 
+                    {
+                        character = result.Character;
+                    }                    
+
+                    switch (result.Command)
+                    {
+                        case Command.InvalidInput:
+                            Console.WriteLine(result.Message);
+                            continue;
+                        case Command.GetItems:
+                            character.PrintItems(); 
+                            continue;
+                        case Command.GetDescription:
+                            character.PrintСharacteristic(); 
+                            continue;
+                        case Command.Start:
+                            character.PrintJson(); 
+                            Console.WriteLine(InputInstructions.AfterStartRules);
+                            continue;
+                        case Command.ShowInfo:
+                            character.PrintJson(); 
+                            continue;
+                        case Command.AddAbility:                            
+                                character.Abilities.Add(result.Ability);
+                                Console.WriteLine($"Способность {result.Ability.Name} добавлена.");                            
+                            continue;
+                    }
+                }
```
1. Команды ShowInfo и AddAbility должны вызываться только после вызова команды Start, Добавь эту логику в парсинг.
2. Логика ShowInfo и AddAbbilty должна вызываться у character который был ранее определен в команде Start, в противовес команды GetItems и GetDescriptions должны вызываться всегда из character в результате парсинга

bekhhh committed:
add some tests

usik-vladislav committed:
Merge branch 'master' into Task9.2
